rekallpy/rekall/interval_set.py

```diff
@@ -0,0 +1,775 @@
+                output. The fields in ``axis`` are overwritten by the grouped
+                values.
+
+        Returns:
+            A new IntervalSet of Intervals grouped by ``axis``, with the full
+            IntervalSet of Intervals in the group in the payload.
+        """
+        def key_fn(intrvl):
+            return (intrvl[axis[0]], intrvl[axis[1]])
+        def merge_fn(key, intervals):
+            new_bounds = output_bounds.copy()
+            new_bounds[axis[0]] = key[0]
+            new_bounds[axis[1]] = key[1]
+            return Interval(new_bounds, intervals)
+        return self.group_by(key_fn, merge_fn)
+
+    def collect_by_interval(self, other, predicate,
+            filter_empty=True, window=None):
+        """Collect intervals in other and nest them under intervals in self.
+
+        For each interval in self, its payload in the output IntervalSet will
+        be all the intervals in other that satisfy the predicate.
+        The payload of each resulting interval will become a pair
+        ``(P, IS)`` where ``P`` is the original payload in self, and ``IS`` is
+        an ``IntervalSet`` containing the nested intervals.
+
+        Note:
+            We do not check the full cross-product of self and other and
+            instead only check pairs that are within ``window`` of each other
+            along the primary axis. See class documentation for more details.
+
+        Args:
+            other (IntervalSet): Set of intervals to collect.
+            predicate: A function that takes one interval in self and one in
+                other and returns a bool.
+            filter_empty (optional): Whether to remove intervals in self if no
+                intervals in other satisfies the predicate with it.
+                If ``False``, such intervals will have payload
+                ``(P, IntervalSet([]))`` where ``P`` is the original payload.
+                Defaults to ``True``.
+            window (optional): Restrict interval pairs to those within
+                window of each other along the primary axis.
+                Defaults to None which means using the default optimization
+                window associated with self. See class Documentation for more
+                detail.
+        
+        Returns:
+            A new IntervalSet of intervals in self but with nested
+            interval set from other in the payload.
+        """
+        def map_output(intrvlself, intrvlothers):
+            intrvls_to_nest = IntervalSet([
+                i for i in intrvlothers if predicate(intrvlself, i)])
+            if not intrvls_to_nest.empty() or not filter_empty:
+                return [Interval(
+                    intrvlself['bounds'].copy(),
+                    (intrvlself['payload'], intrvls_to_nest))]
+            return []
+        return IntervalSet(self._map_with_other_within_primary_axis_window(
+            other, map_output, window))
+
+    def coalesce(self, axis, bounds_merge_op,
```

Comment:
This signature is a bit tricky. The function with the current O(n) implementation only works when `bounds_merge_op` span-merges along `axis` __and__ when `axis` is primary, since `fold` operates on the primary axis.

Given how fragile it is, I suggest providing a simpler function instead that bakes in primary axis and `bounds_merge_op`

Reply:
Good catch - I meant to sort the IntervalSet by `axis` in the fold. Fixed.

At this level of abstraction, there's no `bounds_merge_op` that can work by default, since you don't know what Bounds you're dealing with.